website/moneybirdsynchronization/services.py

```diff
@@ -0,0 +1,257 @@
+import datetime
+
+from django.db.models import OuterRef, Q, Subquery
+from django.template.defaultfilters import date
+
+from moneybirdsynchronization import emails
+from moneybirdsynchronization.administration import Administration
+from moneybirdsynchronization.models import MoneybirdContact, MoneybirdExternalInvoice
+from moneybirdsynchronization.moneybird import MoneybirdAPIService
+
+from events.models.event import Event
+from members.models import Member
+from payments.models import Payment
+from thaliawebsite import settings
+
+
+def get_moneybird_api_service():
+    if (
+        settings.MONEYBIRD_ADMINISTRATION_ID is None
+        or settings.MONEYBIRD_API_KEY is None
+    ):
+        raise RuntimeError("Moneybird API key or administration ID not set")
+    return MoneybirdAPIService(
+        settings.MONEYBIRD_API_KEY, settings.MONEYBIRD_ADMINISTRATION_ID
+    )
+
+
+def push_thaliapay_batch(batch):
+    if settings.MONEYBIRD_SYNC_ENABLED is False:
+        return
+
+    apiservice = get_moneybird_api_service()
+    payments = Payment.objects.filter(batch=batch)
+    tpay_account_id = settings.MONEYBIRD_THALIAPAY_FINANCIAL_ACCOUNT_ID
+    apiservice.link_transaction_to_financial_account(tpay_account_id, payments)
+    batch.pushed_thaliapay_batch.processed = True
+    batch.pushed_thaliapay_batch.save()
+
+
+def update_contact(member):
+    if settings.MONEYBIRD_SYNC_ENABLED is False:
+        return
+
+    apiservice = get_moneybird_api_service()
+    apiservice.update_contact(MoneybirdContact.objects.get_or_create(member=member)[0])
+
+
+def delete_contact(member):
+    if settings.MONEYBIRD_SYNC_ENABLED is False:
+        return
+
+    apiservice = get_moneybird_api_service()
+    apiservice.delete_contact(MoneybirdContact.objects.get(member=member))
+
+
+def create_external_invoice(payment):
+    external_invoice = MoneybirdExternalInvoice(payment=payment)
+    external_invoice.save()
+
+
+def register_event_registration_payment(registration):
+    if settings.MONEYBIRD_SYNC_ENABLED is False:
+        return
+
+    apiservice = get_moneybird_api_service()
+    contact_id = apiservice.get_contact_id_by_customer(registration)
+
+    start_date = date(registration.event.start, "Y-m-d")
+    project_name = f"{registration.event.title} [{start_date}]"
+    project_id = apiservice.get_project_id(project_name)
+
+    invoice_info = apiservice.create_external_sales_info(
+        contact_id, registration.payment, project_id
+    )
+
+    try:
+        response = apiservice.create_external_sales_invoice(invoice_info)
+        registration.payment.moneybird_external_invoice.moneybird_invoice_id = response[
+            "id"
+        ]
+        registration.payment.moneybird_external_invoice.save()
```

Comment:
I think it is better to attach the external invoice to the event registration and not to the payment